### PR TITLE
Optimize log.XXX extensions with T properties arguments.

### PR DESCRIPTION
**Pull request by deniaa**
Double the speed and reduce memory traffic x3 times for the benchmark with nine properties.

This is a benchmark results:

BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19043.1415 (21H1/May2021Update)
Intel Core i7-4771 CPU 3.50GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT

|               Method |       Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |-----------:|---------:|---------:|-------:|------:|------:|----------:|
|    LogWithProperties |   655.8 ns | 12.00 ns | 12.83 ns | 0.1392 |     - |     - |     584 B |
| LogOldWithProperties | 1,024.9 ns | 20.08 ns | 20.62 ns | 0.3643 |     - |     - |   1,528 B |

image
Halved garbage from these methods.

### CODE REVIEW

**Comment by deniaa** [comment]
This pull request requires another one to be accepted first: https://github.com/vostok/commons.collections/pull/10

**Comment by kungasc** [comment]
https://github.com/vostok/logging.abstractions/blob/master/Vostok.Logging.Abstractions/Extensions/LogExtensions_Interpolated.tt#L95

тут тоже плиз переделай на UnsafeSet потом

**Comment by deniaa** [comment]
Новые результаты бенчмарка после переделок после изменений в ImmutableArrayDictionary:

|                         Method |       Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------- |-----------:|---------:|---------:|-------:|------:|------:|----------:|
|        LogWithObjectProperties |   697.9 ns | 13.21 ns | 21.33 ns | 0.1392 |     - |     - |     584 B |
|     LogOldWithObjectProperties | 1,015.3 ns | 19.96 ns | 21.36 ns | 0.3796 |     - |     - |   1,592 B |
|    LogWithDictionaryProperties |   892.4 ns |  8.87 ns |  7.41 ns | 0.3319 |     - |     - |   1,392 B |
| LogOldWithDictionaryProperties | 1,607.0 ns | 30.82 ns | 56.36 ns | 0.5913 |     - |     - |   2,480 B |

File: Vostok.Logging.Abstractions/Extensions/LogExtensions_Interpolated.cs [hunk 1]
```diff
@@ -255,8 +255,8 @@ public void AppendFormatted(object value, [CallerArgumentExpression("value")] st
                 MessageTemplate.Append('{');
                 MessageTemplate.Append(name);
                 MessageTemplate.Append('}');
-
-                Properties = Properties.Set(name, value);
+                
+                Properties.SetUnsafe(name, value, true);
```
Comment (kungasc) [inline Vostok.Logging.Abstractions/Extensions/LogExtensions_Interpolated.cs:259]: может лучше сразу Append?

**Pull request merged**